libs/pika/async_mpi/include/pika/async_mpi/transform_mpi.hpp

```diff
@@ -43,61 +52,94 @@ namespace pika::mpi::experimental {
         {
             detail::add_request_callback(
                 [&op_state](int status) mutable {
+                    using namespace pika::debug::detail;
+                    PIKA_DP(mpi_tran,
+                        debug(str<>("callback_void_suspend_resume"), "stream",
+                            detail::stream_name(op_state.stream)));
                     op_state.ts = {};
-                    PIKA_ASSERT(
-                        std::holds_alternative<Result>(op_state.result));
-                    set_value_request_callback_helper(status,
-                        PIKA_MOVE(op_state.receiver),
-                        PIKA_MOVE(std::get<Result>(op_state.result)));
+                    op_state.status = status;
+
+                    // wake up the suspended thread
+                    {
+                        std::lock_guard lk(op_state.mutex_);
+                        op_state.resume = true;
+                    }
+                    op_state.cond_var_.notify_one();
                 },
-                request, op_state.stream);
+                // we do not need to eagerly check, because it was done earlier
+                request, false, op_state.stream);
         }
 
+        // -----------------------------------------------------------------
+        // can function be invoked with param types + MPI_Request
         template <typename F, typename... Ts>
         inline constexpr bool is_mpi_request_invocable_v =
-            std::is_invocable_v<F,
-                std::add_lvalue_reference_t<std::decay_t<Ts>>..., MPI_Request*>;
+            std::is_invocable_v<F, std::add_lvalue_reference_t<std::decay_t<Ts>>..., MPI_Request*>;
 
+        // -----------------------------------------------------------------
+        // get return type of func(Ts..., MPI_Request)
         template <typename F, typename... Ts>
-        using mpi_request_invoke_result_t =
-            std::decay_t<pika::util::detail::invoke_result_t<F,
-                std::add_lvalue_reference_t<std::decay_t<Ts>>...,
-                MPI_Request*>>;
+        using mpi_request_invoke_result_t = std::decay_t<pika::util::detail::invoke_result_t<F,
+            std::add_lvalue_reference_t<std::decay_t<Ts>>..., MPI_Request*>>;
 
+        // -----------------------------------------------------------------
+        // route calls through an impl layer for ADL resolution
         template <typename Sender, typename F>
         struct transform_mpi_sender_impl
         {
             struct transform_mpi_sender_type;
         };
 
         template <typename Sender, typename F>
-        using transform_mpi_sender = typename transform_mpi_sender_impl<Sender,
-            F>::transform_mpi_sender_type;
+        using transform_mpi_sender = typename transform_mpi_sender_impl<Sender, F>::transform_mpi_sender_type;
 
+        // -----------------------------------------------------------------
```

Comment:
I know this is your preferred style and I prefer comments without these lines, and that this is purely subjective, but can we do without these lines? There's already whitespace to separate things, and the comment also acts as a separator, no? We don't use this style elsewhere.